first_order_model/shrink_util.py

```diff
@@ -0,0 +1,1235 @@
+                          prune_indices)
+                    f_set(node.value.running_mean.detach(),
+                          node.value.running_mean, prune_indices)
+                    f_set(node.value.running_var.detach(),
+                          node.value.running_var, prune_indices)
+
+    return model
+
+
+def get_metrics_loss(metrics_dataloader, lr_size, generator_full,
+                     generator_type):
+    """
+    Computes total loss when running generator_full on metrics_dataloader's all inputs
+    """
+    total_loss = 0
+    with torch.no_grad():
+        for y in metrics_dataloader:
+            y['driving_lr'] = F.interpolate(y['driving'], lr_size)
+            
+            # Inputs need to be moved manually to gpu
+            move_to_gpu(y)
+            losses_generator, metrics_generated = generator_full(
+                y, generator_type)
+            loss_values = [val.mean() for val in losses_generator.values()]
+            loss = sum(loss_values)
+            total_loss += loss.item()
+
+    return total_loss
+
+def get_following_layers_skip_depthwise(following_layers, layer_graph, x):
+    """
+    Specialized version of follo that skips over depthwise convolved layers. Only used when sorting the outputs of a layer for importance.
+
+    """
+    following_layers.append(x.index)
+    if x.type == 'conv' and x.value.groups == 1:
+        return
+    for y in x.after:
+        get_following_layers_skip_depthwise(following_layers, layer_graph, layer_graph[y])
+
+def follow(following_layers, layer_graph, x):
+    """
+    Get any possible layers which are directly impacted by editing x's outputs
+    """
+    following_layers.append(x.index)
+    if x.type == 'conv':
+        return
+    for y in x.after:
+        follow(following_layers, layer_graph, layer_graph[y])
+
+
+def get_first_conv(layer_graph, following_layers):
+    """
+    Gets first convolution in following layers
+    """
+    for layer in following_layers:
+        if layer_graph[layer].type == 'conv' and layer_graph[layer].value.groups == 1:
+            return layer
+
+def compute_deletion(layer_graph,
+                     custom_deletions,
+                     deleted_things,
+                     layer,
+                     sort,
+                     custom,
+                     reason=None):
+    """
+    Given a layer, generate the list of the indexes we need to delete from its following layers
+    """
+    # find all the following layers
+    following_layers = []
+
+    for after_layer in layer_graph[layer].after:
+        follow(following_layers, layer_graph, layer_graph[after_layer])
+
+    # If you delete a part of the previous one, not just all of it then you cannot just say these n are to be deleted
+    # So for each layer you need to figure out what you are deleting from it, then for the next layer figure out what is deleted
+
+    deletions = {}
+    if isinstance(custom, list):
```

Comment:
what is this codepath doing? Not a good idea to overload variables in general, it makes it really hard to follow the code

Reply:
This is potentially confusing but both sort of mean the same thing so they aren't semantically overloaded.

Reply:
I don't like this though. can we write it differently?